### PR TITLE
Start NSSCertExportCLI

### PR DESCRIPTION
**Pull request by cipherboy**
Can be tested with `pki nss-cert-export`

`Signed-off-by: Alexander Scheel <ascheel@redhat.com>`

### CODE REVIEW

File: base/java-tools/src/com/netscape/cmstools/nss/NSSCertExportCLI.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+        if (chain && format.equals("DER")) {
+            throw new Exception("Unable to write chain of DER-encoded certificates; use PEM instead.");
+        }
+
+        MainCLI mainCLI = (MainCLI) getRoot();
+        mainCLI.init();
+
+        ClientConfig clientConfig = mainCLI.getConfig();
+        NSSDatabase nssdb = mainCLI.getNSSDatabase();
```
Comment (edewata) [inline base/java-tools/src/com/netscape/cmstools/nss/NSSCertExportCLI.java:80]: This object doesn't seem to be used.

**Comment by cipherboy** [comment]
I'm closing this in favor of #517 for the time being. We can revisit this CLI at a later point.

File: base/java-tools/src/com/netscape/cmstools/nss/NSSCertExportCLI.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+            for (X509Certificate cert : certs) {
+                byte[] encoded = cert.getEncoded();
+                buffer.append(Cert.HEADER);
+                buffer.append(Utils.base64encodeMultiLine(encoded));
+                buffer.append(Cert.FOOTER);
+            }
+        } else if (format.equals("DER")) {
+            for (X509Certificate cert : certs) {
+                byte[] encoded = cert.getEncoded();
+                buffer.append(encoded);
```
[the remark this answers is not in the document]
Comment (cipherboy) [reply base/java-tools/src/com/netscape/cmstools/nss/NSSCertExportCLI.java:111]: See line 72-ish above:

> ```java
>         if (chain && format.equals("DER")) {
>             throw new Exception("Unable to write chain of DER-encoded certificates; use PEM instead.");
>         }
> ```

[the remark this answers is not in the document]
Comment (cipherboy) [reply base/java-tools/src/com/netscape/cmstools/nss/NSSCertExportCLI.java:111]: (i.e., no, and we only allow exporting a single DER cert).

File: base/java-tools/src/com/netscape/cmstools/nss/NSSCertExportCLI.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+            certs = new X509Certificate[] {
+                (PK11Cert) km.getCertificate(nickname)
+            };
+        }
+
+        StringBuffer buffer = new StringBuffer();
+
+        if (format.equals("RAW")) {
+            for (X509Certificate cert : certs) {
+                buffer.append(certs.toString());
```
[the remark this answers is not in the document]
Comment (cipherboy) [reply base/java-tools/src/com/netscape/cmstools/nss/NSSCertExportCLI.java:99]: Ah sorry, this has been fixed. I needed to cast to `PK11Cert`. That has the pretty `toString()` method. (Also, `certs` was wrong -- should be `cert`).

File: base/java-tools/src/com/netscape/cmstools/nss/NSSCertExportCLI.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+        if (chain) {
+            certs = km.getCertificateChain(nickname);
+        } else {
+            certs = new X509Certificate[] {
+                (PK11Cert) km.getCertificate(nickname)
+            };
+        }
+
+        StringBuffer buffer = new StringBuffer();
```
[the remark this answers is not in the document]
Comment (cipherboy) [reply base/java-tools/src/com/netscape/cmstools/nss/NSSCertExportCLI.java:95]: Done. :)

**Comment by cipherboy** [comment]
Reopened. This will probably work better than trying to fix PKCS#12.

File: base/java-tools/src/com/netscape/cmstools/nss/NSSCertExportCLI.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (edewata) [reply base/java-tools/src/com/netscape/cmstools/nss/NSSCertExportCLI.java]: OK, the code uses a loop, but there's only one certificate. A bit confusing, but should work just fine.

**Comment by edewata** [comment]
@SilleBille I added some patches on top of @cipherboy's patches. Could you take a look? Thanks!

File: base/common/python/pki/nssdb.py [hunk 5]
```diff
@@ -1378,7 +1375,7 @@ def export_cert_from_db(self,
         if output_format:
             cmd.extend(['--format', output_format])
 
-        cmd.extend([full_name, output_file])
+        cmd.extend([nickname, output_file])
```
Comment (SilleBille) [inline base/common/python/pki/nssdb.py:1378]: When you say that the `serverCertNick.conf` already has the prefix added, which file are you referring to? I don't see any processing of nickname for hsm environment

Comment (edewata) [reply base/common/python/pki/nssdb.py:1378]: During installation we store the cert nickname and the token name ([if available](https://github.com/dogtagpki/pki/blob/master/base/server/python/pki/server/instance.py#L519-L520)) into these files:
* [`/etc/pki/<instance>/serverCertNick.conf`](https://github.com/dogtagpki/pki/blob/master/base/server/python/pki/server/instance.py#L522), and
* [`/etc/pki/<instance>/server.xml`](https://github.com/dogtagpki/pki/blob/master/base/server/python/pki/server/instance.py#L532)

During startup the nickname and the token name (if available) will be [reloaded](https://github.com/dogtagpki/pki/blob/master/base/server/python/pki/server/cli/migrate.py#L141) from either of those files using [`PKIInstance.get_sslserver_cert_nickname()`](https://github.com/dogtagpki/pki/blob/master/base/server/python/pki/server/instance.py#L497) as a single value, so it's not necessary to prepend the value with the token name again.

Note that eventually we want to remove `serverCertNick.conf`, but that hasn't been implemented yet.

**Comment by edewata** [comment]
Thanks! I was able to install CA with Luna SA HSM with this patch. Merging..

File: base/common/python/pki/nssdb.py (hunk 5 above)
Comment (cipherboy) [reply base/common/python/pki/nssdb.py:1378]: @edewata Ah, good catch. Thanks!

**Pull request merged**